src/appshell/view/dockwindow/dockwindow.cpp

```diff
@@ -403,15 +477,49 @@ DockToolBarHolder* DockWindow::resolveToolbarDockingHolder(const QPoint& localPo
         newHolder = m_mainToolBarDockingHolder;
     } else if (localPos.y() > centralFrameGeometry.top()
                && localPos.y() < centralFrameGeometry.top() + MAX_DISTANCE_TO_HOLDER) { // page top toolbar holder
-        newHolder = page->holderByLocation(DockBase::DockLocation::Top);
+        newHolder = page->toolBarHolderByLocation(DockBase::DockLocation::Top);
     } else if (localPos.y() < centralFrameGeometry.bottom()) { // page left toolbar holder
         if (localPos.x() < MAX_DISTANCE_TO_HOLDER) {
-            newHolder = page->holderByLocation(DockBase::DockLocation::Left);
+            newHolder = page->toolBarHolderByLocation(DockBase::DockLocation::Left);
         } else if (localPos.x() > mainFrameGeometry.right() - MAX_DISTANCE_TO_HOLDER) { // page right toolbar holder
-            newHolder = page->holderByLocation(DockBase::DockLocation::Right);
+            newHolder = page->toolBarHolderByLocation(DockBase::DockLocation::Right);
         }
     } else if (localPos.y() < mainFrameGeometry.bottom()) { // page bottom toolbar holder
-        newHolder = page->holderByLocation(DockBase::DockLocation::Bottom);
+        newHolder = page->toolBarHolderByLocation(DockBase::DockLocation::Bottom);
+    }
+
+    return newHolder;
+}
+
+DockPanelHolder* DockWindow::resolvePanelDockingHolder(const QPoint& localPos) const
+{
+    const DockPage* page = currentPage();
+    if (!page) {
+        return nullptr;
+    }
+
+    const KDDockWidgets::DockWidgetBase* centralDock = page->centralDock()->dockWidget();
+    if (!centralDock) {
+        return nullptr;
+    }
+
+    QRect centralFrameGeometry = centralDock->frameGeometry();
+    centralFrameGeometry.setTopLeft(m_mainWindow->mapFromGlobal(centralDock->mapToGlobal({ centralDock->x(), centralDock->y() })));
+
+    QRect mainFrameGeometry = m_mainWindow->rect();
+    DockPanelHolder* newHolder = nullptr;
+
+    if (localPos.y() > centralFrameGeometry.top()
+        && localPos.y() < centralFrameGeometry.top() + MAX_DISTANCE_TO_HOLDER) { // page top panel holder
+        newHolder = page->panelHolderByLocation(DockBase::DockLocation::Top);
+    } else if (localPos.y() < centralFrameGeometry.bottom()) { // page left panel holder
+        if (localPos.x() < MAX_DISTANCE_TO_HOLDER) {
```

Comment:
It won't work if there is a toolbar attached to the left.
I think it needs a binding to centralFrameGeometry instead of mainFrameGeometry like for top